File: wordNet/src/main/java/org/sqlunet/wordnet/sql/Synset.java
/*
 * Copyright (c) 2019. Bernard Bou <1313ou@gmail.com>.
 */

package org.sqlunet.wordnet.sql;

import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Synset
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
class Synset extends BasicSynset
{
	static private final String TAG = "Synset";

	/**
	 * Constructor from data
	 *
	 * @param synsetId    synset id
	 * @param definition  definition
	 * @param lexDomainId lexdomain id
	 * @param sample      sample
	 */
	protected Synset(final long synsetId, final String definition, final int lexDomainId, final String sample)
	{
		super(synsetId, definition, lexDomainId, sample);
	}

	/**
	 * Constructor from query for synsets
	 *
	 * @param query query for synsets
	 */
	Synset(@NonNull final SynsetsQueryFromWordId query)
	{
		super(query.getSynsetId(), query.getDefinition(), query.getLexDomainId(), query.getSample());
	}

	/**
	 * Constructor from query for synset
	 *
	 * @param query query for synset
	 */
	Synset(@NonNull final SynsetQuery query)
	{
		super(query.getSynsetId(), query.getDefinition(), query.getLexDomainId(), query.getSample());
	}

	/**
	 * Constructor from query for synsets of a given type
	 *
	 * @param query query for synsets of a given type
	 */
	Synset(@NonNull final SynsetsQueryFromWordIdAndCondition query)
	{
		super(query.getSynsetId(), query.getDefinition(), query.getLexDomainId(), query.getSample());
	}

	/**
	 * Constructor from query for linked synsets
	 *
	 * @param query query for linked synsets
	 */
	Synset(@NonNull final LinksQueryFromSynsetId query)
	{
		super(query.getSynsetId(), query.getDefinition(), query.getLexDomainId(), query.getSamples());
	}

	/**
	 * Constructor from query for synsets linked through a given relation type
	 *
	 * @param query query for synsets linked through a given relation type
	 */
	Synset(@NonNull final LinksQueryFromSynsetIdAndLinkType query)
	{
		super(query.getSynsetId(), query.getDefinition(), query.getLexDomainId(), query.getSamples());
	}

	/**
	 * Get words in the synset as a list
	 *
	 * @param connection connection to the database
	 * @return list of words in synset
	 */
	@Nullable
	public List<Word> getSynsetWords(final SQLiteDatabase connection)
	{
		SynsetWordsQuery query = null;
		List<Word> words = new ArrayList<>();
		try
		{
			query = new SynsetWordsQuery(connection, this.synsetId);
			query.execute();

			while (query.next())
			{
				final String lemma = query.getLemma();
				final long id = query.getId();
				words.add(new Word(lemma, id));
			}
		}
		catch (@NonNull final SQLException e)
		{
			Log.e(TAG, "While querying synset words", e);
			words = null;
		}
		finally
		{
			if (query != null)
			{
				query.release();
			}
		}
		return words;
	}

	/**
	 * Get words in the synset as a string
	 *
	 * @param connection connection to the database
	 * @return list of words in synset as a comma-separated string
	 */
	@NonNull
	public String getSynsetWordsAsString(final SQLiteDatabase connection)
	{
		final StringBuilder sb = new StringBuilder();

		// synset words
		final List<Word> words = getSynsetWords(connection);

		// stringify
		if (words != null)
		{
			for (int i = 0; i < words.size(); i++)
			{
				final Word word = words.get(i);
				final String lemma = word.lemma.replace('_', ' ');
				if (i != 0)
				{
					sb.append(',');
				}
				sb.append(lemma);
			}
		}
		return sb.toString();
	}

	/**
	 * Get the synset's part-of-speech
	 *
	 * @return synset's part-of-speech
	 */
	@NonNull
	public String getPosName()
	{
		return Mapping.getPosName(this.lexDomainId);
	}

	/**
	 * Get the synset's lexdomain name
	 *
	 * @return synset's lexdomain name
	 */
	@NonNull
	public String getLexDomainName()
	{
		return Mapping.getLexDomainName(this.lexDomainId);
	}

	/**
	 * Get synsets linked to the synset
	 *
	 * @param connection connection
	 * @param wordId     word id (for lexical links)
	 * @return list of synsets linked to the synset
	 */
	@Nullable
	public List<Link> getLinks(final SQLiteDatabase connection, final long wordId)
	{
		LinksQueryFromSynsetId query = null;
		List<Link> links = new ArrayList<>();
		try
		{
			query = new LinksQueryFromSynsetId(connection);
			query.setFromSynset(this.synsetId);
			query.setFromWord(wordId);
			query.execute();

			while (query.next())
			{
				final Link link = new Link(query);
				links.add(link);
			}
		}
		catch (@NonNull final SQLException e)
		{
			Log.e(TAG, "While querying links", e);
			links = null;
		}
		finally
		{
			if (query != null)
			{
				query.release();
			}
		}
		return links;
	}

	/**
	 * Get synsets linked to the synset through a given relation type
	 *
	 * @param connection connection
	 * @param wordId     word id (for lexical links)
	 * @param linkType   link type
	 * @return list of synsets linked to the synset through a given relation type
	 */
	@Nullable
	public List<Link> getTypedLinks(final SQLiteDatabase connection, final long wordId, final int linkType)
	{
		LinksQueryFromSynsetIdAndLinkType query = null;
		List<Link> links = new ArrayList<>();
		try
		{
			query = new LinksQueryFromSynsetIdAndLinkType(connection);
			query.setFromSynset(this.synsetId);
			query.setFromWord(wordId);
			query.setLinkType(linkType);
			query.execute();

			while (query.next())
			{
				final Link link = new Link(query);
				links.add(link);
			}
		}
		catch (@NonNull final SQLException e)
		{
			Log.e(TAG, "While querying typed links", e);
			links = null;
		}
		finally
		{
			if (query != null)
			{
				query.release();
			}
		}
		return links;
	}
}

File: wordNet/src/main/java/org/sqlunet/wordnet/sql/Synset.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.wordnet.sql

import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.util.Log

/**
 * Synset
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
open class Synset : BasicSynset {

    val posName: String
        get() = Mapping.getDomainPosName(domainId)

    val domainName: String
        get() = Mapping.getDomainName(domainId)

    /**
     * Constructor from data
     *
     * @param synsetId   synset id
     * @param definition definition
     * @param domainId   domain id
     * @param sample     sample
     */
    constructor(synsetId: Long, definition: String?, domainId: Int, sample: String?) : super(synsetId, definition!!, domainId, sample!!)

    /**
     * Constructor from query for synsets
     *
     * @param query query for synsets
     */
    constructor(query: SynsetsQueryFromWordId) : super(query.synsetId, query.definition, query.domainId, query.sample)

    /**
     * Constructor from query for synset
     *
     * @param query query for synset
     */
    constructor(query: SynsetQuery) : super(query.synsetId, query.definition, query.domainId, query.sample)

    /**
     * Constructor from query for synsets of a given type
     *
     * @param query query for synsets of a given type
     */
    constructor(query: SynsetsQueryFromWordIdAndCondition) : super(query.synsetId, query.definition, query.domainId, query.sample)

    /**
     * Constructor from query for related synsets
     *
     * @param query query for related synsets
     */
    internal constructor(query: RelatedsQueryFromSynsetId) : super(query.synsetId, query.definition, query.domainId, query.samples)

    /**
     * Constructor from query for synsets related through a given relation type id
     *
     * @param query query for synsets related through a given relation type id
     */
    internal constructor(query: RelatedsQueryFromSynsetIdAndRelationId) : super(query.synsetId, query.definition, query.domainId, query.samples)

    /**
     * Get words in the synset as a list
     *
     * @param connection connection to the database
     * @return list of words in synset
     */
    fun getSynsetWords(connection: SQLiteDatabase): List<Word>? {
        try {
            SynsetWordsQuery(connection, synsetId).use {
                it.execute()
                val words: MutableList<Word> = ArrayList()
                while (it.next()) {
                    val word = it.word
                    val id = it.id
                    words.add(Word(word, id))
                }
                return words
            }
        } catch (e: SQLException) {
            Log.e(TAG, "While querying synset words", e)
            return null
        }
    }

    /**
     * Get words in the synset as a string
     *
     * @param connection connection to the database
     * @return list of words in synset as a comma-separated string
     */
    fun getSynsetWordsAsString(connection: SQLiteDatabase): String {
        val sb = StringBuilder()

        // synset words
        val words = getSynsetWords(connection)

        // stringify
        if (words != null) {
            for (i in words.indices) {
                val word = words[i]
                val word2 = word.word.replace('_', ' ')
                if (i != 0) {
                    sb.append(',')
                }
                sb.append(word2)
            }
        }
        return sb.toString()
    }

    /**
     * Get synsets related to the synset
     *
     * @param connection connection
     * @param wordId     word id (for lexical relations)
     * @return list of synsets related to the synset
     */
    open fun getRelateds(connection: SQLiteDatabase, wordId: Long): List<Related>? {
        try {
            RelatedsQueryFromSynsetId(connection).use {
                it.fromSynset = synsetId
                it.fromWord = wordId
                it.execute()
                val relateds: MutableList<Related> = ArrayList()
                while (it.next()) {
                    val related = Related(it)
                    relateds.add(related)
                }
                return relateds
            }
        } catch (e: SQLException) {
            Log.e(TAG, "While querying relateds", e)
            return null
        }
    }

    /**
     * Get synsets related to the synset through a given relation type id
     *
     * @param connection connection
     * @param wordId     word id (for lexical relations)
     * @param relationId relation type id
     * @return list of synsets related to the synset through a given relation type
     */
    fun getTypedRelateds(connection: SQLiteDatabase, wordId: Long, relationId: Int): List<Related>? {
        try {
            RelatedsQueryFromSynsetIdAndRelationId(connection).use {
                it.fromSynset = synsetId
                it.fromWord = wordId
                it.fromRelationId = relationId
                it.execute()
                val relateds: MutableList<Related> = ArrayList()
                while (it.next()) {
                    val related = Related(it)
                    relateds.add(related)
                }
                return relateds
            }
        } catch (e: SQLException) {
            Log.e(TAG, "While querying typed relations", e)
            return null
        }
    }

    companion object {
        private const val TAG = "Synset"
    }
}
